### ISSUE TITLE
chained promises with nesting are executed out of order

### ISSUE THREAD

**Issue by oysterpack** [near/near-sdk-rs#263]
## Example
```rust
let promise_1 = ...
let promise_2 = ...
let promise_3 = ...

// [promise_1 -> promise_2 -> promise_3] - expected execution order
// [promise_2 -> promise_1 -> promise_3] - actual execution order
promise_1.then(promise_2.then(promise_3))

// execution should be the same as:
promise_1.then(promise_2).then(promise_3);
```

I am using:
```toml
[dependencies]
near-sdk = { git = "https://github.com/near/near-sdk-rs",  tag = "2.3.0" }
```

**Comment by evgenykuzyakov** [comment]
It seems there are some issue with the way it getting scheduled in `near-sdk-rs`. I think it does resolve promises on `Drop` which is when they leave the scope.

Needs investigation.

**Comment by austinabell** [comment]
Looking at the code it seems that the 2nd promise would be dropped when replaced with 3. Not quite sure why this would be the case, instead of having a `Vec` of following promises or attaching the future event to the chain of promises, so I'll try to come up with a solution that wouldn't break any other variant

**Comment by nearmax** [comment]
SDK promises replicate behavior of Rust futures, so in this specific example:
```rust
promise_1.then(promise_2.then(promise_3)),
```
`promise_1` and `promise_2` are executed concurrently, potentially on different shards, and there is no guarantee which one is executed first.

However, unlike with Rust futures, in context of NEAR cross-contract calls the following syntax does not make sense:
```rust
promise_1.then(promise_2.then(promise_3))
```
because `x.then(y)` means that `y` is getting execution result of `x` as callback arguments. But is is unclear which promise is getting execution result of `promise_1`, so this syntax should be prohibited, and we should allow only unchained promises to be passed into `.then()` method.

> It seems there are some issue with the way it getting scheduled in `near-sdk-rs`. I think it does resolve promises on `Drop` which is when they leave the scope.
> 
> Needs investigation.

This is WAI. Rust SDK promises implement RAII pattern.

> Looking at the code it seems that the 2nd promise would be dropped when replaced with 3. Not quite sure why this would be the case, instead of having a `Vec` of following promises or attaching the future event to the chain of promises, so I'll try to come up with a solution that wouldn't break any other variant

The correct chaining of the promises should be done `.and()` method, and we should prohibit the above syntax as explained above.

@matklad please share your opinion on it, thank you.

**Comment by austinabell** [comment]
> SDK promises replicate behavior of Rust futures, so in this specific example:
> 
> ```rust
> promise_1.then(promise_2.then(promise_3)),
> ```
> 
> `promise_1` and `promise_2` are executed concurrently, potentially on different shards, and there is no guarantee which one is executed first.

Did you mean to switch the first call to `.and(`? Why would promise_2 not wait on promise_1 if `promise_2.then(promise_3)` awaits on promise_2?

> 
> However, unlike with Rust futures, in context of NEAR cross-contract calls the following syntax does not make sense:
> 
> ```rust
> promise_1.then(promise_2.then(promise_3))
> ```
> 
> because `x.then(y)` means that `y` is getting execution result of `x` as callback arguments. But is is unclear which promise is getting execution result of `promise_1`, so this syntax should be prohibited, and we should allow only unchained promises to be passed into `.then()` method.
> 

Can you maybe explain why `promise_2` wouldn't/couldn't get the execution result of the example above?

So to confirm with this, there can only be `then` chains of two elements, where any `.then(other)` call the `other` must be a single promise? And if there need to be multiple calls done async, they can still be done with `and`?

So visually it can look like:
```
p1 -> p2
p3 -> p4
```

But the following chain is impossible?:

p1 -> p2 -> p3

> The correct chaining of the promises should be done with `.and()` method, and we should prohibit `promise_1.then(promise_2.then(promise_3))` syntax as explained above.
> 
> @matklad please share your opinion on it, thank you.

Maybe if we draw comparisons to

**Comment by bowenwang1996** [comment]
> But the following chain is impossible?:
> 
> ```
> p1 -> p2 -> p3
> ```

This should be written as `p1.then(p2).then(p3)`.

**Comment by austinabell** [comment]
> > But the following chain is impossible?:
> > ```
> > p1 -> p2 -> p3
> > ```
> 
> This should be written as `p1.then(p2).then(p3)`.

Okay I'll look into this soon and try to create  e2e tests, but this was the case where the scheduling was misordered (where `p2` was being dropped and scheduled before the others)

**Comment by matklad** [comment]
I've spend some time looking at it. I feel I don't understand the underlying runtime primitives to suggest an appropriate high-level API. I've look at 

* https://docs.near.org/docs/tutorials/contracts/cross-contract-calls
* https://nomicon.io/RuntimeSpec/Components/BindingsSpec/PromisesAPI.html
* https://github.com/matklad/nearcore/blob/2f577c4f9f5b1c1b2781f624775a511ac867afc7/runtime/near-vm-logic/src/logic.rs#L890

As an example, here are a couple of specific questions I have right now:

*  "what is the difference between `then` and `and` with a single input promise?" 
* Is "callback" a different kind of promise than just a "promise"?

@nearmax, @bowenwang1996 what's the best way to learn about promises?

My current gut feeling is that we are just mixing up two styles of DAG creation API:

* `and` adds a new node and a bunch of edges to it
* `then` links two existing nodes with a new edge

a possible fix is to remove `then`'s argument and make it create a new empty promise.

**Comment by bowenwang1996** [comment]
> "what is the difference between then and and with a single input promise?"

In `then`, you create a promise with a function call action that will act on the result of the previous promise. `and` simply creates a promise that resolves when all the promise dependencies resolve.

**Comment by austinabell** [comment]
Is there any consensus we can reach on this before the next major release? Although I switched it to just panic on runtime in this case mentioned, I still don't understand why in the case:

```
p1.then(p2.then(p3))
```

it can't be assumed that `p2` uses the callback from `p1` and keep the same structure as if the developer wrote:

```
p1.then(p2).then(p3)
```

Maybe an edge case to show why it shouldn't be the same or allowed would help? I think I might be misunderstanding some nuance to this.

If it's still uncertain, it's no rush, we can push for this in a future release where we could possibly change this API so it would fail to compile rather than at runtime in invalid cases.

**Comment by austinabell** [comment]
bump on this, ideally, the stable release comes soon, and would be nice to come to some conclusion on this.

https://github.com/near/near-sdk-rs/pull/410 originally made it so that both cases are handled consistently, which seems logical to me, and I can't really think of a counterpoint, but was changed based on a suggestion to fail at runtime if this other syntax (`p1.then(p2.then(p3))`) is used.

I think either way is better than the current behaviour and would be nice to have some improvement in the next release. I have a preference toward having consistent behaviour and not failing at runtime but would be happy with the alternative as long as there is progress.

**Comment by austinabell** [comment]
Okay so since we are a bit stalled on this, I will suggest a path forward.

I would like to get input from @itegulov and even @ChaoticTempest if you are interested so we can resolve this and pick one of the alternatives as either should be an improvement on the unexpected behaviour of the current impl.

Options for `promise_1.then(promise_2.then(promise_3))`:

1. Make the ordering the same as `promise_1.then(promise_2).then(promise_3)` (roughly https://github.com/near/near-sdk-rs/compare/master...6a26068d74ac1f5b76de762101934e3797347574 but base is out of date). Ordering to me still seems clear and unambiguous that 1 -> 2 -> 3, but I understand this could potentially be interpreted as promise 3 being scheduled only after promise 1 completes.
2. Disallow this pattern at runtime, based on @nearmax's suggestion from https://github.com/near/near-sdk-rs/pull/410 (commit https://github.com/near/near-sdk-rs/pull/410/commits/70d5251136ede5ed7839acfab7f7b959c3e602d0). There is potentially a way to restrict this at compile-time, but would be a highly breaking change and make the API less ergonomic.

Since we are at a bit of a stalemate here, it would be very appreciated if @itegulov drives this forward with a suggestion of which option to take, and ideally, we can have approval from @nearmax and/or @bowenwang1996 that it makes sense before bringing in the change and releasing with stable 4.0

**Comment by itegulov** [comment]
> Make the ordering the same as promise_1.then(promise_2).then(promise_3) (roughly https://github.com/near/near-sdk-rs/compare/master...6a26068d74ac1f5b76de762101934e3797347574 but base is out of date). Ordering to me still seems clear and unambiguous that 1 -> 2 -> 3, but I understand this could potentially be interpreted as promise 3 being scheduled only after promise 1 completes.

While I agree that the ordering is intuitively clear to me as well, I would argue that the argument-passing semantics are not. This kind of pattern actually occurs quite often (lazy evaluations, effectful computations, monads etc), but arguments are always handled explicitly like this: `promise_1.then(|x, y| promise_2.args(x, y).then(...))`. I know that our use case is different and we can't really support a general syntax like this one, but that is also why I think we are better with diverging from this syntax not to confuse people. So, to sum it up, I am in favor of disallowing this pattern at runtime.

> There is potentially a way to restrict this at compile-time, but would be a highly breaking change and make the API less ergonomic.

Just a side note: I can see how this would be a breaking change, but why do you think it would make the API less ergonomic? One horribly overkilly but still a cool way we could prohibit this syntax would be by introducing a Peano arithmetic generic parameter that would "count" how deep the Future instance is (i.e. each Future starts at 0 and every `then` returns a new Future with depth increased by one). Finally, we would only allow `then` accepting instances of `Future<0, T>`. I am sure there are less involved ways to achieve this sort of behaviour as well.

**Comment by austinabell** [comment]
> While I agree that the ordering is intuitively clear to me as well, I would argue that the argument-passing semantics are not. This kind of pattern actually occurs quite often (lazy evaluations, effectful computations, monads etc), but arguments are always handled explicitly like this: `promise_1.then(|x, y| promise_2.args(x, y).then(...))`. I know that our use case is different and we can't really support a general syntax like this one, but that is also why I think we are better with diverging from this syntax not to confuse people. So, to sum it up, I am in favor of disallowing this pattern at runtime.

I'm slightly confused about your example here, as the example you gave is the same pattern you are suggesting to disallow at runtime? Are you saying that since it might not be clear which promise the arguments are used for, it makes sense to disallow this at runtime?

> Just a side note: I can see how this would be a breaking change, but why do you think it would make the API less ergonomic? One horribly overkilly but still a cool way we could prohibit this syntax would be by introducing a Peano arithmetic generic parameter that would "count" how deep the Future instance is (i.e. each Future starts at 0 and every `then` returns a new Future with depth increased by one). Finally, we would only allow `then` accepting instances of `Future<0, T>`. I am sure there are less involved ways to achieve this sort of behaviour as well.

Makes it less ergonomic because a lot of patterns we suggest and have examples of having returned a `Promise` as the function return value. If there is some struct that abstracts above promise, it requires some conversion to a promise. Just by nature of having multiple types with a similar but slightly different API will be harder to use.

--- 

To summarize, just erroring at runtime and accepting that if someone constructs the promises in this pattern they will just have the bad UX of only finding the error when trying to execute through a test (unit or workspace test would catch). If this is the path, we should probably have this clearly documented that this pattern is disallowed and the PR linked I just would update the branch.

**Comment by itegulov** [comment]
> Are you saying that since it might not be clear which promise the arguments are used for, it makes sense to disallow this at runtime?

Precisely, yes. It becomes even weirder if you consider things like this:
```rust
p1.then(p2.and(p3))
```

But thinking on this topic a bit more, I realize that the argument passing semantics are formalizable if you think of `then` (without considering `and`) as something working in a [continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style). So there might be a neat algebraic representation here.

> Makes it less ergonomic because a lot of patterns we suggest and have examples of having returned a Promise as the function return value. If there is some struct that abstracts above promise, it requires some conversion to a promise. Just by nature of having multiple types with a similar but slightly different API will be harder to use.
> 
> I would suggest trying to restrict functionality based on any type of generic to see firsthand the implications and to see how messy it gets quickly.

Yeah, this is probably my naivete speaking here. Doing a small PoC seems like a good way to explore this issue, so give me a few days to try a few things out.

> To summarize, just erroring at runtime and accepting that if someone constructs the promises in this pattern they will just have the bad UX of only finding the error when trying to execute through a test (unit or workspace test would catch). If this is the path, we should probably have this clearly documented that this pattern is disallowed and the PR linked I just would update the branch.

Don't get me wrong, I am generally a big opponent of runtime errors, and I 100% agree with your points in this thread. My last message was more or less a response to being given just two choices (runtime errors or keeping the syntax while changing the ordering).

**Comment by austinabell** [comment]
> > Are you saying that since it might not be clear which promise the arguments are used for, it makes sense to disallow this at runtime?
> 
> Precisely, yes. It becomes even weirder if you consider things like this:
> 
> ```rust
> p1.then(p2.and(p3))
> ```

Note: this pattern is currently disallowed at runtime

> Don't get me wrong, I am generally a big opponent of runtime errors, and I 100% agree with your points in this thread. My last message was more or less a response to being given just two choices (runtime errors or keeping the syntax while changing the ordering).

Sorry, I don't mean to constrain to only these options, I just wanted to list the options proposed in this thread. Feel free to suggest anything else. I just didn't add compile-time errors to this as it would require a full revamp of the promise API and is unlikely to make it in this next stable API, but is something we can consider for the next major version. Want to avoid churn switching this API until we are certain it will be the cleanest in our context.

Erroring at runtime makes sense and is in line with the behaviour of the pattern listed above, I suggest we just go with this and I'll update my PR branch today.

**Comment by ChaoticTempest** [comment]
Maybe I'm repeating a point you guys have already mentioned, but from my brief look through the code, I wonder if we can't just internally create a chain with the correct expected sequence of execution for `p1.then(p2.then(p3))`. The problem seems to stem from the `then` function returning the passed in Promise instead of creating a new one that is internally creating a chain of promises. But maybe I'm missing something here since I'm not sure how the promises completely get scheduled

> > Precisely, yes. It becomes even weirder if you consider things like this:
> > ```rust
> > p1.then(p2.and(p3))
> > ```
> 
> Note: this pattern is currently disallowed at runtime

this is somewhat related as well, but we should probably iron out what the following expected execution looks like for:
```rust
p1.then(p2).and(p3)
```
The docs say p1 and p3 are executed concurrently, but from looking at the promise code, it makes it seem like p2 and p3 would be ran concurrently instead.

**Comment by ChaoticTempest** [comment]
> If they didn't execute concurrently, this would be a bug. What about the code makes it seem like p3 would be executed with p2 and not p1? Are you just saying from the syntax above it seems like this?

Not the syntax, but from how the promise gets created when calling into `and`. For instance, `p1.then(p2)` returns `p2` and then we get `p2.and(p3)` which will execute p2 and p3 concurrently is what seems to be happening. Not sure if this is the correct way to test it, but with viewing it from a mock blockchain test, I get the receipts in `p1 p2 p3` order instead of `p1 p3 p2`.
```rust
        let p1 = Promise::new("bob_near".parse().unwrap()).create_account();
        let p2 = Promise::new("carol_near".parse().unwrap()).create_account();
        let p3 = Promise::new("dave_near".parse().unwrap()).create_account();

        let p = p1.then(p3).and(p2);
        drop(p);

        with_mocked_blockchain(|b| {
            println!("{:?}", b.created_receipts());
        });
```
Would have to test this in workspaces to see what it's actually executing on a node
